shared/lib/accounts/accounts.ts

```diff
@@ -35,46 +50,189 @@ export function getUniqueAccountName(
+        return `Bitcoin Testnet Account ${accountNumber}`;
+      }
+      return `Bitcoin Account ${accountNumber}`;
+    }
+    case SOLANA_WALLET_SNAP_ID: {
+      // Solana accounts should have in their scope the 3 networks
+      // mainnet, testnet, and devnet. Therefore, we can use this name
+      // for all 3 networks.
+      return `Solana Account ${accountNumber}`;
     }
+    default:
+      return defaultSnapAccountName;
+  }
+}
+
+///: BEGIN:ONLY_INCLUDE_IF(solana,bitcoin)
+export type WalletSnapOptions = {
+  scope?: CaipChainId;
+  derivationPath?: DiscoveredAccount['derivationPath'];
+  entropySource?: string;
+  accountNameSuggestion?: string;
+};
+
+export type WalletSnapClient = {
+  getSnapId(): SnapId;
+
+  createAccount(
+    options: WalletSnapOptions,
+    internalOptions?: SnapKeyringInternalOptions,
+  ): Promise<KeyringAccount>;
+
+  getNextAvailableAccountName(
+    options?: SnapAccountNameOptions,
+  ): Promise<string>;
+};
+
+export type MultichainWalletSnapClientMessenger = Messenger<
+  | SnapControllerHandleRequest
+  | AccountsControllerGetNextAvailableAccountNameAction,
+  never
+>;
+
+export class MultichainWalletSnapClient implements WalletSnapClient {
+  #snapId: SUPPORTED_WALLET_SNAP_ID;
+
+  #snapKeyring: SnapKeyring;
+
+  #scopes: CaipChainId[];
+
+  #client: KeyringInternalSnapClient;
 
-    await Promise.allSettled(
-      discovered.map(async (discoveredAccount) => {
-        const options = {
-          derivationPath: discoveredAccount.derivationPath,
-          entropySource,
-        };
-
-        // TODO: Use `withKeyring` instead of using the keyring directly.
-        await snapKeyring.createAccount(snapId, options, {
-          displayConfirmation: false,
-          displayAccountNameSuggestion: false,
-          setSelectedAccount: false,
-        });
+  #messenger: MultichainWalletSnapClientMessenger;
+
+  constructor(
+    snapId: SUPPORTED_WALLET_SNAP_ID,
+    scopes: CaipChainId[],
+    snapKeyring: SnapKeyring,
+    messenger: MultichainWalletSnapClientMessenger,
+  ) {
+    this.#snapId = snapId;
+    this.#snapKeyring = snapKeyring;
+
+    this.#scopes = scopes;
+
+    this.#messenger = messenger;
+
+    this.#client = new KeyringInternalSnapClient({
+      snapId,
+      // @ts-expect-error TODO: Resolve mismatch between base-controller versions.
+      messenger: messenger.getRestricted({
+        name: 'KeyringInternalSnapClient',
+        allowedActions: ['SnapController:handleRequest'],
+        allowedEvents: [],
       }),
+    });
+  }
+
+  getSnapId(): SnapId {
+    return this.#snapId;
+  }
+
+  async createAccount(
+    options: WalletSnapOptions,
+    internalOptions?: SnapKeyringInternalOptions,
+  ): Promise<KeyringAccount> {
+    // TODO: Use `withKeyring` instead of using the keyring directly.
+    return await this.#snapKeyring.createAccount(
+      this.#snapId,
+      options,
+      internalOptions,
+    );
+  }
```

Comment:
To make this even more seamless, I thought we could automatically set the `accountNameSuggestion` on `options` if the caller does not give us anything? This way just calling `createAccount` would just create the account and name it properly (based on our naming rule).

Any opinion on this?

Reply:
Decided to actually implement this here: https://github.com/MetaMask/metamask-extension/pull/32122/commits/1cd397628cd806284b49bf2ae12616ed56996330